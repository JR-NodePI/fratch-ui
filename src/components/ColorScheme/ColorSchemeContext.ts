import { createContext } from 'react';

import type { ColorScheme } from './ColorScheme';
import { COLOR_SCHEMES } from './colorSchemeConstants';

const ColorSchemeContext = createContext<
  [colorScheme: ColorScheme, setColorScheme: (data: ColorScheme) => void]
>([COLOR_SCHEMES.LIGHT, () => {}]);

ColorSchemeContext.displayName = 'ColorSchemeContext';

export default ColorSchemeContext;
