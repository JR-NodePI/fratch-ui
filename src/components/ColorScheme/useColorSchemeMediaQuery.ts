import { useEffect } from 'react';

import type { ColorSchemeOutput } from './ColorScheme';
import { COLOR_SCHEMES } from './colorSchemeConstants';
import usePersistedColorScheme from './usePersistedColorScheme';

const COLOR_SCHEME_DATA_ATTR_NAME = 'data-fratch_color_scheme';
const COLOR_SCHEME_MEDIA_QUERY = '(prefers-color-scheme: dark)';

export default function useColorSchemeMediaQuery(): ColorSchemeOutput {
  const initialColorScheme = window.matchMedia(COLOR_SCHEME_MEDIA_QUERY).matches
    ? COLOR_SCHEMES.DARK
    : COLOR_SCHEMES.LIGHT;
  const [colorScheme, setColorScheme] =
    usePersistedColorScheme(initialColorScheme);

  useEffect(() => {
    document.body.setAttribute(COLOR_SCHEME_DATA_ATTR_NAME, colorScheme);
  }, [colorScheme]);

  useEffect(() => {
    const handler = (e: MediaQueryListEvent): void => {
      setColorScheme(e.matches ? COLOR_SCHEMES.DARK : COLOR_SCHEMES.LIGHT);
    };

    window
      .matchMedia(COLOR_SCHEME_MEDIA_QUERY)
      .addEventListener('change', handler);

    return () => {
      window
        .matchMedia(COLOR_SCHEME_MEDIA_QUERY)
        .removeEventListener('change', handler);
    };
  }, [setColorScheme]);

  return [colorScheme, setColorScheme];
}
