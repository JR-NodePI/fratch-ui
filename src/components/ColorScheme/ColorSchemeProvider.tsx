import { ColorSchemeProviderProps } from './ColorScheme';
import ColorSchemeContext from './ColorSchemeContext';
import useColorSchemeMediaQuery from './useColorSchemeMediaQuery';

export default function ColorSchemeProvider({
  children,
}: ColorSchemeProviderProps): JSX.Element {
  const [colorScheme, setColorScheme] = useColorSchemeMediaQuery();

  return (
    <ColorSchemeContext.Provider value={[colorScheme, setColorScheme]}>
      {children}
    </ColorSchemeContext.Provider>
  );
}
