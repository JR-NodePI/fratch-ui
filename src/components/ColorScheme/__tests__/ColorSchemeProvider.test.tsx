import { useContext } from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import ColorSchemeContext from '../ColorSchemeContext';
import ColorSchemeProvider from '../ColorSchemeProvider';
import useColorSchemeMediaQuery from '../useColorSchemeMediaQuery';

vi.mock('../useColorSchemeMediaQuery');

describe('ColorSchemeSwitcher', () => {
  const setColorScheme = vi.fn();

  beforeEach(() => {
    (
      useColorSchemeMediaQuery as jest.MockedFunction<
        typeof useColorSchemeMediaQuery
      >
    ).mockReturnValue(['light', setColorScheme]);
  });

  const setup = () => {
    const MockChildComponent = () => {
      const [colorScheme, setColorScheme] = useContext(ColorSchemeContext);
      const handleClick = () => {
        setColorScheme('dark');
      };
      return (
        <>
          {JSON.stringify({ colorScheme })}
          <button onClick={handleClick}>CHANGE</button>
        </>
      );
    };

    return render(
      <ColorSchemeProvider>
        <MockChildComponent />
      </ColorSchemeProvider>
    );
  };

  afterEach(() => {
    vi.resetAllMocks();
    setColorScheme.mockClear();
  });

  it('should render properly', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();

    expect(useColorSchemeMediaQuery).toHaveBeenCalledTimes(1);
  });

  it('should set a new color schema', async () => {
    setup();

    await userEvent.click(screen.getByRole('button'));

    expect(setColorScheme).toHaveBeenCalledTimes(1);
    expect(setColorScheme).toHaveBeenCalledWith('dark');
  });
});
