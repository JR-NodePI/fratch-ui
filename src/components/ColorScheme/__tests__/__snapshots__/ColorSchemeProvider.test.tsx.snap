// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColorSchemeSwitcher > should render properly 1`] = `
<div>
  {"colorScheme":"light"}
  <button>
    CHANGE
  </button>
</div>
`;
